src/implementation/js/node/int.ts

```diff
@@ -0,0 +1,196 @@
+          this.readUInt16LE(out);
+        } else {
+          this.readUInt16BE(out);
+        }
+        break;
+      }
+
+      case 3: {
+        if (this.ref.littleEndian) {
+          this.readUInt24LE(out);
+        } else {
+          this.readUInt24BE(out);
+        }
+        break;
+      }
+
+      case 4: {
+        if (this.ref.littleEndian) {
+          this.readUInt32LE(out);
+        } else {
+          this.readUInt32BE(out);
+        }
+        break;
+      }
+    }
+
+    this.tailTo(out, this.ref.otherwise!);
+  }
+
+  private readInt8(out: string[]) {
+    const ctx = this.compilation;
+    const index = ctx.stateField(this.ref.field);
+
+    out.push(`${index} = (${ctx.bufArg()}[${ctx.offArg()}] & 2 ** 7) * 0x1fffffe;`)
+  }
+
+  private readUInt8(out: string[]) {
+    const ctx = this.compilation;
+    const index = ctx.stateField(this.ref.field);
+
+    out.push(`${index} = ${ctx.bufArg()}[${ctx.offArg()}];`)
+  }
+
+  private readUInt16LE(out: string[]) {
+    const ctx = this.compilation;
+    const index = ctx.stateField(this.ref.field);
+
+    switch (this.ref.byteOffset) {
+      case 0: {
+        out.push(`${index} = ${ctx.bufArg()}[${ctx.offArg()}];`)
+        break;
+      }
+
+      case 1: {
+        out.push(`${index} += ${ctx.bufArg()}[${ctx.offArg()}] * 2 ** 8;`);
+        break;
+      }
+    }
+  }
+
+  private readUInt24LE(out: string[]) {
+    const ctx = this.compilation;
+    const index = ctx.stateField(this.ref.field);
+
+    switch (this.ref.byteOffset) {
+      case 0: {
+        out.push(`${index} = ${ctx.bufArg()}[${ctx.offArg()}];`)
+        break;
+      }
+
+      case 1: {
+        out.push(`${index} += ${ctx.bufArg()}[${ctx.offArg()}] * 2 ** 8;`);
+        break;
+      }
+
+      case 2: {
+        out.push(`${index} += ${ctx.bufArg()}[${ctx.offArg()}] * 2 ** 16;`);
+        break;
+      }
+    }
+  }
+
+  private readUInt32LE(out: string[]) {
+    const ctx = this.compilation;
+    const index = ctx.stateField(this.ref.field);
+
+    switch (this.ref.byteOffset) {
+      case 0: {
+        out.push(`${index} = ${ctx.bufArg()}[${ctx.offArg()}];`)
+        break;
+      }
+
+      case 1: {
+        out.push(`${index} += ${ctx.bufArg()}[${ctx.offArg()}] * 2 ** 8;`);
+        break;
+      }
+
+      case 2: {
+        out.push(`${index} += ${ctx.bufArg()}[${ctx.offArg()}] * 2 ** 16;`);
+        break;
+      }
+
+      case 3: {
+        out.push(`${index} += ${ctx.bufArg()}[${ctx.offArg()}] * 2 ** 24;`);
+        break;
+      }
+    }
+  }
+
+  private readUInt16BE(out: string[]) {
+    const ctx = this.compilation;
+    const index = ctx.stateField(this.ref.field);
+
+    switch (this.ref.byteOffset) {
+      case 0: {
+        out.push(`${index} = ${ctx.bufArg()}[${ctx.offArg()}] * 2 ** 8;`);
+        break;
+      }
+
+      case 1: {
+        out.push(`${index} += ${ctx.bufArg()}[${ctx.offArg()}];`)
+        break;
+      }
+    }
+  }
+
+  private readUInt24BE(out: string[]) {
```

Comment:
Sounds like it could be generalized a bit. There is no runtime savings from having separate methods for `16`, `24`, etc bits. These functions are executed only at compile time, right?

Reply:
Yes. My thinking was to keep these separated for readability - having all the logic in a single huge `if/elseif/else` statement made this really hard to understand. 🤷‍♂ Let's see how this looks like once all the other operations are in place too.